File: specs/typespec/global/global.go
package global

import (
	"vetchium-api-server.typespec/common"
)

// Structs
type Region struct {
	RegionCode string `json:"region_code"`
	RegionName string `json:"region_name"`
}

type SupportedLanguage struct {
	LanguageCode string `json:"language_code"`
	LanguageName string `json:"language_name"`
	NativeName   string `json:"native_name"`
	IsDefault    bool   `json:"is_default"`
}

type CheckDomainRequest struct {
	Domain common.DomainName `json:"domain"`
}

func (r CheckDomainRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type CheckDomainResponse struct {
	IsApproved bool `json:"is_approved"`
}

type GetRegionsResponse struct {
	Regions []Region `json:"regions"`
}

type GetSupportedLanguagesResponse struct {
	Languages []SupportedLanguage `json:"languages"`
}

File: specs/typespec/global/global.ts
import {
	type DomainName,
	type ValidationError,
	newValidationError,
	validateDomainName,
} from "../common/common";

export type { DomainName, ValidationError };

// Interfaces
export interface Region {
	region_code: string;
	region_name: string;
}

export interface SupportedLanguage {
	language_code: string;
	language_name: string;
	native_name: string;
	is_default: boolean;
}

export interface CheckDomainRequest {
	domain: DomainName;
}

export interface CheckDomainResponse {
	is_approved: boolean;
}

export interface GetRegionsResponse {
	regions: Region[];
}

export interface GetSupportedLanguagesResponse {
	languages: SupportedLanguage[];
}

// Request validators

export function validateCheckDomainRequest(
	request: CheckDomainRequest,
): ValidationError[] {
	const errs: ValidationError[] = [];

	const domainErr = validateDomainName(request.domain);
	if (domainErr) {
		errs.push(newValidationError("domain", domainErr));
	}

	return errs;
}
